File: finanze/application/use_cases/get_position.py
from datetime import date

from application.ports.entity_port import EntityPort
from application.ports.position_port import PositionPort
from dateutil.relativedelta import relativedelta
from domain.global_position import EntitiesPosition, PositionQueryRequest, ProductType
from domain.use_cases.get_position import GetPosition


def _calculate_next_loan_payment_date(loan):
    today = date.today()

    if loan.maturity <= today:
        return None

    anchor = loan.creation

    if anchor > today:
        candidate = anchor
    else:
        candidate = anchor
        while candidate <= today:
            candidate = candidate + relativedelta(months=1)

    if candidate > loan.maturity:
        return None

    return candidate


def _enrich_loans(position):
    if ProductType.LOAN not in position.products:
        return

    loans = position.products[ProductType.LOAN]
    if not loans or not loans.entries:
        return

    for loan in loans.entries:
        if loan.next_payment_date is None:
            loan.next_payment_date = _calculate_next_loan_payment_date(loan)


def _enrich_data(data: dict):
    for entity_id, position in data.items():
        if position is None:
            continue
        _enrich_loans(position)


class GetPositionImpl(GetPosition):
    def __init__(self, position_port: PositionPort, entity_port: EntityPort):
        self._position_port = position_port
        self._entity_port = entity_port

    async def execute(self, query: PositionQueryRequest) -> EntitiesPosition:
        excluded_entities = [
            e.id for e in await self._entity_port.get_disabled_entities()
        ]

        query = PositionQueryRequest(
            entities=query.entities, excluded_entities=excluded_entities
        )
        global_position_by_entity = (
            await self._position_port.get_last_grouped_by_entity(query)
        )

        global_position_by_entity = {
            str(entity.id): position
            for entity, position in global_position_by_entity.items()
        }

        _enrich_data(global_position_by_entity)

        return EntitiesPosition(global_position_by_entity)

File: mobile/src/application/usecases/getPosition.ts
import {
  EntitiesPosition,
  ProductType,
  PositionQueryRequest,
  GlobalPosition,
} from "@/domain"
import { PositionPort, EntityPort } from "../ports"
import { GetPosition } from "@/domain/usecases"

function daysInMonth(year: number, monthZeroBased: number): number {
  return new Date(year, monthZeroBased + 1, 0).getDate()
}

function formatLocalDate(date: Date): string {
  const y = date.getFullYear()
  const m = String(date.getMonth() + 1).padStart(2, "0")
  const d = String(date.getDate()).padStart(2, "0")
  return `${y}-${m}-${d}`
}

function calculateNextLoanPaymentDate(): string {
  // Backend uses: today + relativedelta(months=1), then keeps today's day.
  // We clamp the day to the target month length to avoid invalid dates.
  const today = new Date()
  today.setHours(0, 0, 0, 0)

  const targetYear = today.getFullYear()
  const targetMonth = today.getMonth() + 1

  const nextMonthDate = new Date(targetYear, targetMonth, 1)
  const dim = daysInMonth(nextMonthDate.getFullYear(), nextMonthDate.getMonth())
  const day = Math.min(today.getDate(), dim)

  const next = new Date(
    nextMonthDate.getFullYear(),
    nextMonthDate.getMonth(),
    day,
  )
  return formatLocalDate(next)
}

function enrichLoans(position: any): void {
  if (!position.products || !(ProductType.LOAN in position.products)) {
    return
  }

  const loans = position.products[ProductType.LOAN]
  if (!loans?.entries) {
    return
  }

  for (const loan of loans.entries) {
    if (loan.nextPaymentDate == null) {
      loan.nextPaymentDate = calculateNextLoanPaymentDate()
    }
  }
}

function enrichData(data: Record<string, GlobalPosition>): void {
  for (const [entityId, position] of Object.entries(data)) {
    if (position == null) continue
    enrichLoans(position)
  }
}

export class GetPositionImpl implements GetPosition {
  constructor(
    private positionPort: PositionPort,
    private entityPort: EntityPort,
  ) {}

  async execute(query?: PositionQueryRequest): Promise<EntitiesPosition> {
    const disabledEntities = await this.entityPort.getDisabledEntities()
    const excludedEntities = disabledEntities
      .map(e => e.id)
      .filter((id): id is string => Boolean(id))

    const fullQuery: PositionQueryRequest = {
      entities: query?.entities,
      excludedEntities,
    }

    const positionsDataByEntity =
      await this.positionPort.getLastGroupedByEntity(fullQuery)

    const positionsData: Record<string, GlobalPosition> = {}
    for (const [entity, position] of positionsDataByEntity.entries()) {
      if (entity.id) {
        positionsData[entity.id] = position
      }
    }

    enrichData(positionsData)

    return { positions: positionsData }
  }
}
